include/sycl_wrappers.h

```diff
@@ -6,32 +6,48 @@
 
 namespace celerity {
 
-namespace detail {
-
-	struct read_only_tag_t {};
-	struct read_write_tag_t {};
-	struct write_only_tag_t {};
-	struct read_only_host_task_tag_t {};
-	struct read_write_host_task_tag_t {};
-	struct write_only_host_task_tag_t {};
-
-	using memory_scope_ut = std::underlying_type_t<cl::sycl::memory_scope>;
-
-} // namespace detail
-
 using access_mode = cl::sycl::access::mode;
 
 enum class target {
 	device,
 	host_task,
 };
 
-inline constexpr detail::read_only_tag_t read_only;
-inline constexpr detail::read_write_tag_t read_write;
-inline constexpr detail::write_only_tag_t write_only;
-inline constexpr detail::read_only_host_task_tag_t read_only_host_task;
-inline constexpr detail::read_write_host_task_tag_t read_write_host_task;
-inline constexpr detail::write_only_host_task_tag_t write_only_host_task;
+} // namespace celerity
+
+namespace celerity::detail {
+
+template <access_mode Mode, access_mode NoInitMode, target Target>
+struct access_tag {};
```

Comment:
Very elegant solution!

Reply:
Thanks. It does remove the helpful `Invalid access tag, expecting one of celerity::{read_only,read_write,write_only}[_host_task]` error messages though.